File: spec/parser_spec.rb
require "json"
require "serpapi_challenge/parser"

RSpec.describe SerpapiChallenge::Parser do
  let(:parser) { described_class.new }

  it "parses the Van Gogh page into the expected structure" do
    input = File.expand_path("../files/van-gogh-paintings.html", __dir__)
    skip "Fixture not present" unless File.exist?(input)

    result = parser.parse_file(input)
    expect(result).to be_a(Hash)
    expect(result["artworks"]).to be_an(Array)
    expect(result["artworks"].length).to be > 0

    first = result["artworks"].first
    expect(first.keys.sort).to eq(%w[extensions image link name].sort)
    expect(first["name"]).to be_a(String)
    expect(first["link"]).to start_with("http")
    expect(first["extensions"]).to be_a(Array)
  end

  it "keeps only embedded (data URL) thumbnails" do
    input = File.expand_path("../files/van-gogh-paintings.html", __dir__)
    skip "Fixture not present" unless File.exist?(input)

    result = parser.parse_file(input)
    images = result["artworks"].map { |h| h["image"] }.compact
    expect(images.all? { |src| src.start_with?("data:image/") }).to be(true)
  end

  context "additional fixtures" do
    Dir[File.expand_path("fixtures/*.html", __dir__)].sort.each do |fixture|
      it "parses #{File.basename(fixture)}" do
        html = File.read(fixture, encoding: "UTF-8")
        result = parser.parse_html(html)
        expect(result["artworks"]).to be_an(Array)
        expect(result["artworks"].length).to be > 0
      end
    end

    it "skips if none added" do
      fixtures = Dir[File.expand_path("fixtures/*.html", __dir__)]
      skip "Add 2 similar SERP pages to spec/fixtures to run robustness tests" if fixtures.empty?
    end
  end
end

File: src/__test__/parser.spec.js
const path = require('path');
const { parse } = require('../index');

describe('Google search results page parser', () => {
  let artworks;
  [
    'american-novels.html',
    'ruby-rail-book.html',
    'van-gogh-paintings.html',
  ].map((file, index) => {
    describe(`Parse file: ${file}`, () => {
      beforeAll(async () => {
        artworks = await parse(
          path.join(
            `file://${path.join(__dirname)}`,
            '../../files/',
            file
          )
        );
      });

      test('Should have artworks parsed', () => {
        expect(artworks.length > 0).toBeTruthy();
      });

      describe('Artworks results', () => {
        test('Has a string name', () => {
          expect(
            typeof artworks[0].name === 'string' && artworks[0].name.length > 0
          ).toBeTruthy();
        });
        test('Has a google search link', () => {
          expect(
            artworks[0].link.startsWith('https://www.google.com/search?')
          ).toBeTruthy();
        });
        test('Has a array of extensions or null', () => {
          expect(
            !artworks[0].extensions || Array.isArray(artworks[0].extensions)
          ).toBeTruthy();
        });
        test('Has a image src or null', () => {
          expect(
            artworks[0].image == null ||
              artworks[0].image.startsWith('data:image/')
          ).toBeTruthy();
        });
      });
    });
  });
});
